### PR TITLE
useBaseField: Refactor to typescript

### PR DESCRIPTION
**Pull request by manzoorwanijk**
## What?
Converts `useBaseField` to TypeScript.

## Why?
Part of the @wordpress/components's TypeScript migration #35744.

## How?
- Converts `useBaseField` to TypeScript
- Converts Unit Tests to TypeScript

## Testing Instructions
- Run `npm run dev` and confirm no typing errors

**Note**

GitHub doesn't show the diff for `hook.ts` even though I used `git mv` to rename the file. So, you can check https://github.com/WordPress/gutenberg/commit/99a4f7b46131b3d361272f23b6b627eb8a66e17a to see the changes to the file.

### CODE REVIEW

File: packages/components/src/base-field/hook.ts [hunk 1]
```diff
@@ -0,0 +1,97 @@
...
+ * 				{ ...inputProps }
+ * 				disabled={ disabled }
+ * 			/>
+ * 			{affix}
+ * 		</View>
+ * 	);
+ * }
+ * ```
+ *
+ * @param {WordPressComponentProps< UseBaseFieldProps, 'div' >} props
```
Comment (ciampo) [inline packages/components/src/base-field/hook.ts:60]: No need to defined types twice — the TypeScript notation should be enough

```suggestion
 * @param  props
```

File: packages/components/src/base-field/hook.ts [hunk 2]
```diff
@@ -0,0 +1,97 @@
...
+				isSubtle && styles.subtle,
+				hasError && styles.error,
+				isInline && styles.inline,
+				className
+			),
+		[ className, controlGroupStyles, cx, hasError, isInline, isSubtle ]
+	);
+
+	return {
+		...useFlex( { children: null, ...flexProps, className: classes } ),
```
Comment (ciampo) [inline packages/components/src/base-field/hook.ts:93]: Why are we overriding `children: null` ?

File: packages/components/src/base-field/types.ts [hunk 3]
```diff
@@ -0,0 +1,23 @@
+/**
+ * Internal dependencies
+ */
+import type { FlexProps } from '../flex/types';
+
+export interface UseBaseFieldProps extends Partial< FlexProps > {
```
Comment (ciampo) [inline packages/components/src/base-field/types.ts:6]: We usually use `type` instead of `interface`, would you mind switching to that?

Comment (ciampo) [inline packages/components/src/base-field/types.ts:6]: Can we more specific to which `FlexProps` should be picked here? Any reason why we're not allowing all props from `FlexProps` ?

File: packages/components/src/base-field/test/index.tsx [hunk 4]
```diff
@@ -8,13 +8,17 @@ import { render } from '@testing-library/react';
  */
 import { useBaseField } from '../index';
 import { View } from '../../view';
+import type { WordPressComponentProps } from '../../ui/context';
+import type { UseBaseFieldProps } from '../types';
 
-const TestField = ( props ) => {
+const TestField = (
+	props: WordPressComponentProps< UseBaseFieldProps, 'div' >
+) => {
```
Comment (ciampo) [inline packages/components/src/base-field/test/index.tsx:16]: `WordPressComponentProps` is not necessary here

```suggestion
const TestField = ( props: UseBaseFieldProps ) => {
```

File: packages/components/src/base-field/test/index.tsx [hunk 5]
```diff
@@ -52,10 +56,18 @@ describe( 'base field', () => {
...
-				const result = useBaseField( { disabled, defaultValue } );
+				let result = useBaseField( { disabled, defaultValue } );
+
+				expect( result.disabled ).toBe( disabled );
+				expect( result.defaultValue ).toBe( defaultValue );
+
+				disabled = false;
+				defaultValue = 'something-else';
+
+				result = useBaseField( { disabled, defaultValue } );
```
Comment (ciampo) [inline packages/components/src/base-field/test/index.tsx:70]: Let's rebase first, and then we can see if this test can be written without this "wrapping in the component" hack

File: packages/components/src/base-field/types.ts [hunk 6]
```diff
@@ -0,0 +1,23 @@
...
+	 */
+	disabled?: boolean;
+	/**
+	 * Renders as an inline element (layout).
+	 */
+	isInline?: boolean;
+	/**
+	 * Renders a subtle variant.
+	 */
+	isSubtle?: boolean;
```
Comment (ciampo) [inline packages/components/src/base-field/types.ts:22]: `hasError`, `isInline` and `isSubtle` are missing the `@default` value in their JSDocs

**Review by ciampo — COMMENTED** [review]
Thank you for working on this, and sorry for the late review!

Apart from the inline comments, we may also need to update the README to make sure it's aligned to the types.

Could you also rebase this PR and solve conflicts? Thank you!

**Comment by manzoorwanijk** [comment]
I could finish this and many other TS-related improvements around the time this PR was created but now I may not get more time in the near future to work on this. Someone else may take it over in their fork.

**Comment by ciampo** [comment]
Thank you for the heads-up! And feel free to reach out again anytime you'd like to contribute in the future!

**Pull request closed without merging**